### DISCUSSION TITLE
Selecting on one component of a composite index

### DISCUSSION THREAD

**Discussion by bigbenchrob** [isar/isar#371] [Help] [answered]
The documentation led me to believe that, using a two-element compound index, ids in the index would be organized by the first component of the index, then sorted within that by the second component of the index. Selecting on the first component would give me the second sort 'for free'. Instead, I don't seem to be able to select based on the first component of the index at all, with or without adding a selection on the second component  My test class looks like this:

```dart
part 'isar_test.g.dart';

@Collection(accessor: 'test')
class IsarTest {
  IsarTest( this.id, this.sequence, this.citySingle,  this.cityComposite, this.name,);

  @Id()
  final int id;

  final int sequence;

  @Index(type: IndexType.hash)
  final String citySingle;

  @Index(composite: [CompositeIndex('sequence')])
  final String cityComposite;

  final String name;
  }
}
```
I populate the database with a series of records, the id incrementing as each record is added and the "sequence" field decrementing (so I can distinguish sorting based on this field from the natural order of records in the db, or the id field).

If I select on the citySingle index, it works as expected (the correct selection is made, with the objects returned either in the order in which they were added to the database or by the primary id index field (I can't tell which):

 
```dart
final result = await isarTest
        .where()
        .citySingleEqualTo('Toronto')
        .findAll();

    for (final item in result) {
      print(item.toString());
    }
```
```dart
flutter:
    id:            3
    sequence:      107
    citySingle:    Toronto
    cityComposite: Toronto
    name:          Ruth
flutter:
    id:            6
    sequence:      104
    citySingle:    Toronto
    cityComposite: Toronto
    name:          Helen
flutter:
    id:            9
    sequence:      101
    citySingle:    Toronto
    cityComposite: Toronto
    name:          Emma
```
I thought that if I used the compound index, the records would be returned selected for city (cityComposite == 'Toronto') and sorted by the sequence field. There is no reason to specify a particular sequence value or range -- I just want to take advantage of the sorting. Isar allows me to write a query like this:

```dart
final result =
        await isarTest.where().cityCompositeEqualTo('Toronto').findAll();
```
...but the result set is empty. Querying:

```dart
    final result = await isarTest
        .where()
        .cityCompositeEqualToSequenceGreaterThan('Toronto', 1)
        .findAll();
```
...returns all the records in the database, in the order I would expect the index to be structured: sorted alphabetically by city, and within each city category, numerically by sequence.

I can accomplish what I need to do using a query like:

```dart
    final result = await isarTest
        .where()
        .citySingleEqualTo('Toronto')
        .sortBySequence()
        .findAll();
```
...but I'd like to know what it is I'm not understanding from the documentation.

Thanks in advance!

**Comment by simc** [comment, marked as answer]
Thanks for the detailed bug report. You are completely right and I think this is an issue. Opened one here: #372

**Reply by bigbenchrob** [reply]
Many thanks for your prompt attention and for a really elegant database solution!
